File: gym_management/gym_management/page/trainer/trainer.js
// Copyright (c) 2026, Ransford Borketey and contributors
// For license information, please see license.txt
//
// Desk "Trainer" page - the Desk-side equivalent of the website Trainer
// dashboard (www/gym-portal/trainer.*). A Gym Trainer signs into the Desk
// and works from here instead of the website portal; Gym Staff/Gym
// Manager/System Manager get a Trainer picker so they can look at (and
// check members in/out for) any trainer, not just their own.

frappe.pages['trainer'].on_page_load = function (wrapper) {
	let page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Trainer'),
		single_column: true,
	});

	new GymTrainerPage(page);
};

class GymTrainerPage {
	constructor(page) {
		this.page = page;
		this.selected_trainer = null;

		this.$wrapper = $(`
			<div class="gtp-wrapper" style="padding: 0 5px;">
				<div class="gtp-toolbar" style="display:flex; align-items:center; gap: 12px; margin-bottom: 15px; flex-wrap: wrap;"></div>
				<div class="gtp-empty text-muted" style="display:none; padding: 30px 0;"></div>
				<div class="gtp-body" style="display:none;">
					<h4>${__('My Class Schedule')}</h4>
					<div class="gtp-classes"></div>

					<h4 style="margin-top: 25px;">${__('Upcoming PT Sessions')}</h4>
					<div class="gtp-pt-sessions"></div>

					<h4 style="margin-top: 25px;">${__('My Members')}</h4>
					<div class="gtp-members"></div>
				</div>
			</div>
		`).appendTo(this.page.body);

		this.refresh();
	}

	call(method, args) {
		return frappe.call({
			method: `gym_management.gym_management.page.trainer.trainer.${method}`,
			args: args || {},
		});
	}

	refresh() {
		this.call('get_dashboard', { trainer: this.selected_trainer }).then((r) => {
			this.data = r.message || {};
			this.render_toolbar();
			if (!this.data.trainer) {
				this.$wrapper.find('.gtp-body').hide();
				this.$wrapper.find('.gtp-empty').text(
					__('No Trainer profile is linked to your account. Ask a manager to link it under Trainer > User.')
				).show();
				return;
			}
			this.$wrapper.find('.gtp-empty').hide();
			this.$wrapper.find('.gtp-body').show();
			this.render_classes();
			this.render_pt_sessions();
			this.render_members();
		});
	}

	render_toolbar() {
		const $toolbar = this.$wrapper.find('.gtp-toolbar').empty();
		const t = this.data.trainer;

		if (t) {
			$(`<div><strong>${frappe.utils.escape_html(t.trainer_name)}</strong>
				<span class="text-muted">&nbsp;&mdash;&nbsp;${frappe.utils.escape_html(t.status || '')}</span></div>`)
				.appendTo($toolbar);
		}

		if (this.data.can_pick) {
			const $select = $('<select class="form-control" style="width: 220px;"></select>').appendTo($toolbar);
			$select.append(`<option value="">${__('Select a trainer...')}</option>`);
			this.call('get_trainers').then((r) => {
				(r.message || []).forEach((row) => {
					const opt = $(`<option value="${row.name}">${frappe.utils.escape_html(row.trainer_name)}</option>`);
					if (t && row.name === t.name) opt.prop('selected', true);
					$select.append(opt);
				});
			});
			$select.on('change', () => {
				this.selected_trainer = $select.val() || null;
				this.refresh();
			});
		}
	}

	render_classes() {
		const $el = this.$wrapper.find('.gtp-classes').empty();
		const rows = this.data.classes || [];
		if (!rows.length) {
			$el.append(`<div class="text-muted">${__('No classes assigned to you.')}</div>`);
			return;
		}
		const $table = $(`<table class="table table-bordered"><thead><tr>
			<th>${__('Class')}</th><th>${__('Day')}</th><th>${__('Time')}</th><th>${__('Room')}</th><th>${__('Capacity')}</th>
		</tr></thead><tbody></tbody></table>`).appendTo($el);
		const $tbody = $table.find('tbody');
		rows.forEach((c) => {
			$tbody.append(`<tr>
				<td>${frappe.utils.escape_html(c.class_type || '')}</td>
				<td>${frappe.utils.escape_html(c.day_of_week || '')}</td>
				<td>${frappe.utils.escape_html(c.start_time || '')} - ${frappe.utils.escape_html(c.end_time || '')}</td>
				<td>${frappe.utils.escape_html(c.room || '-')}</td>
				<td>${c.capacity != null ? c.capacity : '-'}</td>
			</tr>`);
		});
	}

	render_pt_sessions() {
		const $el = this.$wrapper.find('.gtp-pt-sessions').empty();
		const rows = this.data.pt_sessions || [];
		if (!rows.length) {
			$el.append(`<div class="text-muted">${__('No upcoming PT sessions.')}</div>`);
			return;
		}
		const $table = $(`<table class="table table-bordered"><thead><tr>
			<th>${__('Member')}</th><th>${__('Date')}</th><th>${__('Time')}</th><th>${__('Duration')}</th>
		</tr></thead><tbody></tbody></table>`).appendTo($el);
		const $tbody = $table.find('tbody');
		rows.forEach((s) => {
			$tbody.append(`<tr>
				<td><a href="/app/gym-member/${encodeURIComponent(s.member)}">${frappe.utils.escape_html(s.member)}</a></td>
				<td>${frappe.utils.escape_html(s.session_date || '')}</td>
				<td>${frappe.utils.escape_html(s.start_time || '-')}</td>
				<td>${s.duration_minutes || '-'} min</td>
			</tr>`);
		});
	}

	render_members() {
		const $el = this.$wrapper.find('.gtp-members').empty();
		const rows = this.data.members || [];
		const checked_in = new Set(this.data.checked_in_members || []);
		if (!rows.length) {
			$el.append(`<div class="text-muted">${__('No members assigned to you yet.')}</div>`);
			return;
		}
		const $table = $(`<table class="table table-bordered"><thead><tr>
			<th>${__('Member')}</th><th>${__('Status')}</th><th>${__('Phone')}</th><th></th>
		</tr></thead><tbody></tbody></table>`).appendTo($el);
		const $tbody = $table.find('tbody');
		rows.forEach((m) => {
			const is_in = checked_in.has(m.name);
			const $tr = $(`<tr>
				<td><a href="/app/gym-member/${encodeURIComponent(m.name)}">${frappe.utils.escape_html(m.member_name)}</a></td>
				<td>${frappe.utils.escape_html(m.membership_status || '')}</td>
				<td>${frappe.utils.escape_html(m.phone || '-')}</td>
				<td></td>
			</tr>`);
			const $btn = $(
				`<button class="btn btn-xs ${is_in ? 'btn-default' : 'btn-primary'}">${is_in ? __('Check Out') : __('Check In')}</button>`
			);
			$btn.on('click', () => {
				$btn.prop('disabled', true);
				this.call(is_in ? 'check_out' : 'check_in', { member: m.name })
					.then(() => {
						frappe.show_alert({ message: is_in ? __('Checked out.') : __('Checked in.'), indicator: 'green' });
						this.refresh();
					})
					.catch(() => {
						$btn.prop('disabled', false);
					});
			});
			$tr.find('td').last().append($btn);
			$tbody.append($tr);
		});
	}
}

File: gym_management/gym_management/page/trainer/trainer.py
# Copyright (c) 2026, Ransford Borketey and contributors
# For license information, please see license.txt
"""Server side for the Desk "Trainer" page (gym_management.page.trainer).

Desk equivalent of the website Trainer dashboard (www/gym-portal/trainer.*):
a Gym Trainer logs into the Desk (not the website) and works from this page
instead of /gym-portal/trainer. Unlike the website portal - built for
website users with no DocType permissions, where every action does its own
ownership check before an ignore_permissions=True write - a Desk user
reaching this page already holds a real role (Gym Trainer / Gym Staff / Gym
Manager / System Manager, see trainer.json's "roles") with its own
DocType-level permissions (see generate.py's trainer_rw() / staff_rw() /
manager_full()), so the mutating calls below (check_in/check_out) insert
and save normally rather than forcing ignore_permissions.

A plain Gym Trainer only ever sees their own linked Trainer record - the
"trainer" argument is ignored for them and silently resolved to their own
profile. Gym Staff/Gym Manager/System Manager may pass any Trainer name to
look at (front-desk staff covering for a trainer, or a manager checking a
member in), which is what the page's own Trainer picker is for.
"""

import frappe
from frappe import _
from frappe.utils import get_datetime

ALLOWED_ROLES = {"System Manager", "Gym Manager", "Gym Staff", "Gym Trainer"}
STAFF_ROLES = {"System Manager", "Gym Manager", "Gym Staff"}


def _check_access():
	if not ALLOWED_ROLES & set(frappe.get_roles()):
		frappe.throw(_("Not permitted to use the Trainer page."), frappe.PermissionError)


def _own_trainer():
	return frappe.db.get_value("Trainer", {"user": frappe.session.user}, "name")


def _resolve_trainer(trainer=None):
	"""A plain Gym Trainer is always pinned to their own record, even if a
	different `trainer` was passed in - only staff-level roles may look at
	someone else's."""
	roles = set(frappe.get_roles())
	own = _own_trainer()
	if roles & STAFF_ROLES:
		return trainer or own
	if not own:
		frappe.throw(_("No Trainer profile is linked to your account."), frappe.PermissionError)
	return own


@frappe.whitelist()
def get_trainers():
	"""Trainer picker options - staff-level roles only; a plain Gym Trainer
	never needs to pick, they only ever see their own record."""
	_check_access()
	if not set(frappe.get_roles()) & STAFF_ROLES:
		return []
	return frappe.get_all(
		"Trainer", fields=["name", "trainer_name"], order_by="trainer_name asc", ignore_permissions=True
	)


@frappe.whitelist()
def get_dashboard(trainer=None):
	_check_access()
	trainer_name = _resolve_trainer(trainer)
	if not trainer_name:
		return {"trainer": None, "can_pick": bool(set(frappe.get_roles()) & STAFF_ROLES)}

	trainer_doc = frappe.get_all(
		"Trainer",
		filters={"name": trainer_name},
		fields=["name", "trainer_name", "department", "status"],
		ignore_permissions=True,
	)
	if not trainer_doc:
		frappe.throw(_("Trainer {0} not found.").format(trainer_name))
	trainer_doc = trainer_doc[0]

	classes = frappe.get_all(
		"Class Schedule",
		filters={"trainer": trainer_name, "is_active": 1},
		fields=["name", "class_type", "day_of_week", "start_time", "end_time", "room", "capacity"],
		order_by="day_of_week asc, start_time asc",
		ignore_permissions=True,
	)

	pt_sessions = frappe.get_all(
		"PT Session",
		filters={"trainer": trainer_name, "status": "Scheduled"},
		fields=["name", "member", "session_date", "start_time", "duration_minutes", "pt_package_purchase"],
		order_by="session_date asc, start_time asc",
		limit_page_length=30,
		ignore_permissions=True,
	)

	member_names = set()
	for row in frappe.get_all(
		"Gym Membership",
		filters={"trainer": trainer_name, "status": "Active"},
		fields=["member"],
		ignore_permissions=True,
	):
		member_names.add(row.member)
	for row in frappe.get_all(
		"Assigned Workout Plan", filters={"assigned_by": trainer_name}, fields=["member"], ignore_permissions=True
	):
		member_names.add(row.member)
	for row in frappe.get_all(
		"Assigned Diet Plan", filters={"assigned_by": trainer_name}, fields=["member"], ignore_permissions=True
	):
		member_names.add(row.member)

	members = []
	if member_names:
		members = frappe.get_all(
			"Gym Member",
			filters={"name": ["in", list(member_names)]},
			fields=["name", "member_name", "membership_status", "phone"],
			order_by="member_name asc",
			ignore_permissions=True,
		)

	open_checkins = frappe.get_all(
		"Gym Attendance",
		filters={"check_out_time": ["is", "not set"]},
		fields=["member"],
		ignore_permissions=True,
	)
	checked_in_members = {row.member for row in open_checkins}

	return {
		"trainer": trainer_doc,
		"can_pick": bool(set(frappe.get_roles()) & STAFF_ROLES),
		"classes": classes,
		"pt_sessions": pt_sessions,
		"members": members,
		"checked_in_members": list(checked_in_members),
	}


@frappe.whitelist()
def check_in(member):
	_check_access()
	if not frappe.db.exists("Gym Member", member):
		frappe.throw(_("Gym Member {0} not found.").format(member))
	att = frappe.new_doc("Gym Attendance")
	att.member = member
	att.check_in_time = get_datetime()
	att.checked_in_by = frappe.session.user
	att.insert()
	return {"name": att.name}


@frappe.whitelist()
def check_out(member):
	_check_access()
	open_att = frappe.get_all(
		"Gym Attendance",
		filters={"member": member, "check_out_time": ["is", "not set"]},
		fields=["name"],
		order_by="check_in_time desc",
		limit_page_length=1,
	)
	if not open_att:
		frappe.throw(_("No open check-in found for this member."))
	att = frappe.get_doc("Gym Attendance", open_att[0].name)
	att.check_out_time = get_datetime()
	att.save()
	return {"name": att.name}
